[AOTI] Support ReinterpretView in abi mode

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114169

https://github.com/pytorch/pytorch/pull/113967 added support for
ReinterpretView but it turnes out we codegen it differently in abi
compat mode. This PR adds support for abi compat mode as well.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

oulgen committed:
[AOTI] Support ReinterpretView in abi mode

https://github.com/pytorch/pytorch/pull/113967 added support for
ReinterpretView but it turnes out we codegen it differently in abi
compat mode. This PR adds support for abi compat mode as well.

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -60,7 +60,9 @@ def is_aligned(
         if isinstance(x, TensorArg):
-            if x.buffer.startswith("reinterpret_tensor"):
+            if x.buffer.startswith("reinterpret_tensor") or x.buffer.startswith(
+                "RAIIAtenTensorHandle"
```
This looks like too general? Many things can start with `RAIIAtenTensorHandle`: do we want to return `False` on all of them?

oulgen replied:
Well all we know at this point is that the name is `RAIIAtenTensorHandle(tmp_tensor_handle_0)`. Previously other RAIIAtenTensorHandle would have failed because there would be no matching tensor in `self.name_to_node`, so returning False on all of them is the right direction.

The longer term fix would be to remove these prefixes from tensor names, even better not use tensor names to do these comparisons.

chenyang78 replied:
I share a similar concern that checking ```RAIIAtenTensorHandle``` in a general purpose utility function ```config_of``` seems to be too aggressive. I am wondering if we could lift the check to its caller in this particular case, i.e.:

https://github.com/pytorch/pytorch/blob/36869463e08a8aeb2ce74aa54920b2f8859bc123/torch/_inductor/codegen/wrapper.py#L893-L899

where we could check if the kernel is of ReinterpretView, then we just set False to the "configs" field of ```triton_meta```. Otherwise, we call the general ```config_of``` function. In this way, we would not pollute the general utility function.

jansel commented:
Is `RAIIAtenTensorHandle` always the same `reinterpret_tensor`?  Can you give an example of what the buffer is?

oulgen commented:
https://github.com/pytorch/pytorch/blob/77f16eb00cb8b4d4ead76f284d0de0c6e8f63764/torch/_inductor/codegen/wrapper.py#L2004-L2090

in this code for `codegen_reinterpret_view`, we either generate `RAIIAtenTensorHandle` or `reinterpret_tensor`. 

On my test case, the emitted code is P887066630

oulgen committed:
Update on "[AOTI] Support ReinterpretView in abi mode"


https://github.com/pytorch/pytorch/pull/113967 added support for
ReinterpretView but it turnes out we codegen it differently in abi
compat mode. This PR adds support for abi compat mode as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aakhundov reviewed (approved):
Looks more clean and specific now. Thanks!

oulgen commented:
@pytorchbot merge
